### PR TITLE
complete task4.1

### PR DESCRIPTION
**Pull request by dyamorph**
Add functionality that allows add, change, delete products and services.
Add functionality for showing a product card with additional services.
Add a functional test for product controller.

### CODE REVIEW

File: app/Http/Controllers/ClientController.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+use App\Models\Product;
+use App\Models\Service;
+use App\Repositories\Interfaces\ProductRepositoryInterface;
+
+class ClientController extends Controller
+{
+    private ProductRepositoryInterface $productRepository;
+
+    public function __construct(ProductRepositoryInterface $productRepository)
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ClientController.php]:
```
public function __construct(
    protected readonly ProductRepositoryInterface $productRepository
)
```

File: app/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+namespace App\Http\Controllers;
+
+use App\Models\Product;
+use App\Repositories\Interfaces\ProductRepositoryInterface;
+
+class ProductController extends Controller
+{
+    private ProductRepositoryInterface $productRepository;
+
+    public function __construct(ProductRepositoryInterface $productRepository)
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: same

File: app/Http/Controllers/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        return view('admin.products.index', compact('products'));
+    }
+
+    public function create()
+    {
+        return view('admin.products.create');
+    }
+
+    public function store()
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: `ProductStoreRequest`

php artisan make:request `ProductStoreRequest`

Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: `public function store(ProductStoreRequest $request)`

File: app/Http/Controllers/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $product = request()->validate([
+            'title' => '',
+            'description' => '',
+            'manufacturer' => '',
+            'release_date' => '',
+            'price' => '',
+        ]);
+
+        $this->productRepository->store($product);
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]:
```
$this->productRepository->store([
'title' => $request->title,
            'description' => '',
            'manufacturer' => '',
            'release_date' => '',
            'price' => '',
]);
```

File: app/Http/Controllers/ProductController.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        return redirect()->route('product.index');
+    }
+
+    public function show(Product $product)
+    {
+        $product = $this->productRepository->find($product->id);
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: unnecessary

File: app/Http/Controllers/ProductController.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        return view('admin.products.show', compact('product'));
+    }
+
+    public function edit(Product $product)
+    {
+        $product = $this->productRepository->find($product->id);
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: same

File: app/Http/Controllers/ProductController.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+        return view('admin.products.edit', compact('product'));
+    }
+
+    public function update(Product $product)
+    {
+        $data = request()->validate([
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: same

File: app/Http/Controllers/ProductController.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        ]);
+
+        $this->productRepository->update($data, $product->id);
+
+        return redirect()->route('product.show', $product->id);
+    }
+
+    public function destroy(Product $product)
+    {
+        $this->productRepository->destroy($product->id);
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: `$this->productRepository->destroy($product);`

File: app/Http/Controllers/ServiceController.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace App\Http\Controllers;
+
+use App\Models\Service;
+
+class ServiceController extends Controller
+{
+    public function index()
+    {
+        $services = Service::all();
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ServiceController.php]: implement repository for services

File: app/Models/Product.php [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Product extends Model
+{
+    use HasFactory;
+    protected $guarded = [];
```
Comment (vladislavzubetcs) [inline app/Models/Product.php]: fill fillable array

Comment (vladislavzubetcs) [reply app/Models/Product.php]: release_date -> add to `protected casts = [];`

File: app/Models/Service.php [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Service extends Model
+{
+    use HasFactory;
+    protected $guarded = [];
```
Comment (vladislavzubetcs) [inline app/Models/Service.php]: same

File: app/Repositories/Interfaces/ProductRepositoryInterface.php [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace App\Repositories\Interfaces;
+
+interface ProductRepositoryInterface
+{
+    public function all();
+    public function store($data);
```
Comment (vladislavzubetcs) [inline app/Repositories/Interfaces/ProductRepositoryInterface.php]: type for $data

File: app/Repositories/Interfaces/ProductRepositoryInterface.php [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+    public function find($id);
```
Comment (vladislavzubetcs) [inline app/Repositories/Interfaces/ProductRepositoryInterface.php]: same

File: app/Repositories/Interfaces/ProductRepositoryInterface.php (hunk 12 above)
Comment (vladislavzubetcs) [reply app/Repositories/Interfaces/ProductRepositoryInterface.php]: and return types

File: app/Repositories/ProductRepository.php [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public function find($id)
+    {
+        return Product::find($id);
+    }
+
+    public function update($data, $id)
+    {
+        $product = Product::where('id', $id)->first();
```
Comment (vladislavzubetcs) [inline app/Repositories/ProductRepository.php]: $product->update([
...
]);

File: app/Repositories/ProductRepository.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        return Product::create($data);
+    }
+
+    public function find($id)
+    {
+        return Product::find($id);
+    }
+
+    public function update($data, $id)
```
Comment (vladislavzubetcs) [inline app/Repositories/ProductRepository.php]: public function update(Product $product, ...): Product

File: database/migrations/2023_08_25_080402_create_products_table.php [hunk 16]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_080402_create_products_table.php]: return new class () extends Migration

File: database/migrations/2023_08_25_080402_create_products_table.php [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('products', function (Blueprint $table) {
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_080402_create_products_table.php]: Schema::create('products', static function (Blueprint $table)

File: database/migrations/2023_08_25_080402_create_products_table.php [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+            $table->id();
+            $table->string('title');
+            $table->text('description');
+            $table->string('manufacturer');
+            $table->text('release_date');
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_080402_create_products_table.php]: $table->date('release_date');

File: database/migrations/2023_08_25_081216_create_services_table.php [hunk 19]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_081216_create_services_table.php]: same

File: database/migrations/2023_08_25_081216_create_services_table.php [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('services', function (Blueprint $table) {
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_081216_create_services_table.php]: same

File: resources/css/app.css [hunk 21]
```diff
@@ -0,0 +1,3 @@
+body {
```
Comment (vladislavzubetcs) [inline resources/css/app.css]: ?

File: resources/views/admin/layouts/base.blade.php [hunk 22]
```diff
@@ -0,0 +1,42 @@
+<!doctype html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport"
+          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-4bw+/aepP/YC94hEpVNVgiZdgIC5+VKNBQNGCHeKRQN+PtmoHDEXuppvnDJzQIu9" crossorigin="anonymous">
```
Comment (vladislavzubetcs) [inline resources/views/admin/layouts/base.blade.php]: ?

File: resources/views/admin/layouts/base.blade.php [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    <link rel="stylesheet" href="{{ asset('css/app.css') }}">
```
Comment (vladislavzubetcs) [inline resources/views/admin/layouts/base.blade.php]: `Vite::asset()`

File: resources/views/admin/services/edit.blade.php [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+            <div class="mb-3">
+                <label for="title" class="form-label">Title</label>
+                <input name="title" type="text" class="form-control" id="title" placeholder="Warranty service" value="{{ $service->title }}">
+            </div>
+            <div class="mb-3">
+                <label for="description" class="form-label">Terms, days</label>
+                <input name="terms" type="text" class="form-control" id="terms" placeholder="2" value="{{ $service->terms }}">
+            </div>
+            <div class="mb-3">
+                <label for="price" class="form-label">Price, BYN</label>
```
Comment (vladislavzubetcs) [inline resources/views/admin/services/edit.blade.php]: `{{ __('Price, BYN'); }}`

File: resources/views/admin/services/edit.blade.php [hunk 25]
```diff
@@ -0,0 +1,24 @@
+@extends('admin.layouts.base')
+
+@section('content')
+    <section class="px-3">
+        <h3 class="mb-4 mt-4">Edit service</h3>
```
Comment (vladislavzubetcs) [inline resources/views/admin/services/edit.blade.php]: {{ __('Edit service') }}

Comment (vladislavzubetcs) [reply resources/views/admin/services/edit.blade.php]: lang -> en/front.php
`'edit_service' => 'Edit service'`,

{{ __('front.edit_service') }}

File: .DS_Store
Comment (vladislavzubetcs) [inline .DS_Store:1]: add to gitignore

File: database/migrations/2023_08_25_080402_create_products_table.php [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+            $table->integer('price');
```
Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_080402_create_products_table.php:20]: add 
`$table->enum('price', Currencies::class)->default(Currencies::BYN->value);`

add to `app\Enums`

`enum Currencies {}` with implementation

Comment (vladislavzubetcs) [inline database/migrations/2023_08_25_080402_create_products_table.php:18]: add cast to model

```
protected casts = [
    'currency' => Currencies::class,
];
```

File: routes/web.php [hunk 27]
```diff
@@ -12,7 +15,28 @@
...
+Route::redirect('/admin', '/admin/products');
 
-Route::get('/', function () {
-    return view('welcome');
+Route::get('/', [ClientController::class, 'index'])->name('client.index');
+Route::get('/{product}', [ClientController::class, 'show'])->name('client.show');
+
+Route::prefix('admin')->group(function () {
+    Route::resource('services', ServiceController::class)->names([
+        'index' => 'service.index',
```
Comment (vladislavzubetcs) [inline routes/web.php]: unnecessary

File: routes/web.php [hunk 28]
```diff
@@ -12,7 +15,28 @@
...
+        'create' => 'service.create',
+        'store' => 'service.store',
+        'edit' => 'service.edit',
+        'update' => 'service.update',
+        'destroy' => 'service.destroy',
+    ]);
+
+    Route::resource('products', ProductController::class)->names([
+        'index' => 'product.index',
```
Comment (vladislavzubetcs) [inline routes/web.php]: same

**Review by vladislavzubetcs — CHANGES_REQUESTED** [review]
good job!

File: app/Data/ProductData.php [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+
+class ProductData extends Data
+{
+    public function __construct(
+        public string $title,
+        public string $description,
+        public string $manufacturer,
+        public string $release_date,
+        public int $price,
+    ) {}
```
Comment (vladislavzubetcs) [inline app/Data/ProductData.php:34]: add validation rules

File: app/Data/ServiceData.php [hunk 30]
```diff
@@ -0,0 +1,14 @@
...
+
+use Spatie\LaravelData\Data;
+
+class ServiceData extends Data
+{
+    public function __construct(
+        public string $title,
+        public int $terms,
+        public int $price,
+    ) {}
```
Comment (vladislavzubetcs) [inline app/Data/ServiceData.php:25]: same

File: app/Enums/Currencies.php [hunk 31]
```diff
@@ -0,0 +1,13 @@
+<?php
+
```
Comment (vladislavzubetcs) [inline app/Enums/Currencies.php:2]: declare(strict_types=1);

File: app/Data/ProductData.php [hunk 32]
```diff
@@ -0,0 +1,16 @@
+<?php
+
```
Comment (vladislavzubetcs) [inline app/Data/ProductData.php:2]: same

File: app/Data/ServiceData.php [hunk 33]
```diff
@@ -0,0 +1,14 @@
+<?php
+
```
Comment (vladislavzubetcs) [inline app/Data/ServiceData.php:2]: same

File: app/Http/Controllers/ClientController.php [hunk 34]
```diff
@@ -1,28 +1,28 @@
 <?php
 
+declare(strict_types=1);
+
 namespace App\Http\Controllers;
 
 use App\Models\Product;
 use App\Models\Service;
 use App\Repositories\Interfaces\ProductRepositoryInterface;
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ClientController.php]: as ProductRepository;

File: app/Http/Controllers/ClientController.php [hunk 35]
```diff
@@ -1,28 +1,28 @@
...
+    public function __construct(
+        protected readonly ProductRepositoryInterface $productRepository
+    ) {}
 
-    public function index()
+    public function index(): View
     {
         $products = $this->productRepository->all();
 
         return view('client.index', compact('products'));
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ClientController.php]:
```
return view('client.index', [
    'products' => $products,
]);
```

Comment (vladislavzubetcs) [reply app/Http/Controllers/ClientController.php]: return view('client.index', [
    'products' => $this->productRepository->all(),
]);

File: app/Http/Controllers/ProductController.php [hunk 36]
```diff
@@ -1,79 +1,61 @@
...
 class ProductController extends Controller
 {
-    private ProductRepositoryInterface $productRepository;
-
-    public function __construct(ProductRepositoryInterface $productRepository)
-    {
-        $this->productRepository = $productRepository;
-    }
+    public function __construct(
+        protected readonly ProductRepositoryInterface $productRepository
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: same

File: app/Http/Controllers/ProductController.php [hunk 37]
```diff
@@ -1,79 +1,61 @@
...
-
         return view('admin.products.show', compact('product'));
     }
 
-    public function edit(Product $product)
+    public function edit(Product $product): View
     {
-        $product = $this->productRepository->find($product->id);
-
         return view('admin.products.edit', compact('product'));
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ProductController.php]: same

File: app/Http/Controllers/ServiceController.php [hunk 38]
```diff
@@ -1,58 +1,56 @@
...
 use App\Models\Service;
+use App\Repositories\Interfaces\ServiceRepositoryInterface;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\View\View;
 
 class ServiceController extends Controller
 {
-    public function index()
+    public function __construct(
+        protected readonly ServiceRepositoryInterface $serviceRepository
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ServiceController.php]: same

File: app/Http/Controllers/ServiceController.php [hunk 39]
```diff
@@ -1,58 +1,56 @@
...
+    ) {}
+
+    public function index(): View
     {
-        $services = Service::all();
+        $services = $this->serviceRepository->all();
 
         return view('admin.services.index', compact('services'));
```
Comment (vladislavzubetcs) [inline app/Http/Controllers/ServiceController.php]: same

File: app/Http/Requests/Product/ProductStoreRequest.php [hunk 40]
```diff
@@ -0,0 +1,34 @@
...
+
+declare(strict_types=1);
+
+namespace App\Http\Requests\Product;
+
+use Illuminate\Http\Request;
+
+class ProductStoreRequest extends Request
+{
+    /**
```
Comment (vladislavzubetcs) [inline app/Http/Requests/Product/ProductStoreRequest.php]: delete comments

File: app/Http/Requests/Product/ProductUpdateRequest.php [hunk 41]
```diff
@@ -0,0 +1,26 @@
...
+        return false;
+    }
+
+    public function rules(): array
+    {
+        return [
+            'title' => 'required',
+            'description' => 'required',
+            'manufacturer' => 'required',
+            'release_date' => 'required|date',
```
Comment (vladislavzubetcs) [inline app/Http/Requests/Product/ProductUpdateRequest.php]: ['required', 'date'],

File: app/Models/Product.php [hunk 42]
```diff
@@ -1,13 +1,27 @@
...
 
+use App\Enums\Currencies;
 use Illuminate\Database\Eloquent\Factories\HasFactory;
 use Illuminate\Database\Eloquent\Model;
 
 class Product extends Model
 {
     use HasFactory;
-    protected $guarded = [];
+    protected $fillable = [
```
Comment (vladislavzubetcs) [inline app/Models/Product.php:15]: empty space before

File: app/Models/Product.php [hunk 43]
```diff
@@ -1,13 +1,27 @@
...
+        'title',
+        'description',
+        'manufacturer',
+        'release_date',
+        'price',
+        'currency'
+    ];
     protected $table = 'products';
+    protected $casts = [
```
Comment (vladislavzubetcs) [inline app/Models/Product.php:28]: same

File: app/Providers/RepositoryServiceProvider.php [hunk 44]
```diff
@@ -0,0 +1,29 @@
...
+
+use App\Repositories\Interfaces\ProductRepositoryInterface;
+use App\Repositories\Interfaces\ServiceRepositoryInterface;
+use App\Repositories\ProductRepository;
+use App\Repositories\ServiceRepository;
+use Illuminate\Support\ServiceProvider;
+
+class RepositoryServiceProvider extends ServiceProvider
+{
+    /**
```
Comment (vladislavzubetcs) [inline app/Providers/RepositoryServiceProvider.php]: same

File: app/Providers/RepositoryServiceProvider.php [hunk 45]
```diff
@@ -0,0 +1,29 @@
...
+     * Register services.
+     */
+    public function register(): void
+    {
+        $this->app->bind(ProductRepositoryInterface::class, ProductRepository::class);
```
Comment (vladislavzubetcs) [inline app/Providers/RepositoryServiceProvider.php]: $this->app->bind(
ProductRepositoryInterface::class,
ProductRepository::class
);

File: app/Providers/RepositoryServiceProvider.php [hunk 46]
```diff
@@ -0,0 +1,29 @@
...
+        $this->app->bind(ServiceRepositoryInterface::class, ServiceRepository::class);
+    }
+
+    /**
+     * Bootstrap services.
+     */
+    public function boot(): void
```
Comment (vladislavzubetcs) [inline app/Providers/RepositoryServiceProvider.php]: delete

File: app/Repositories/Interfaces/ProductRepositoryInterface.php [hunk 47]
```diff
@@ -1,12 +1,16 @@
...
+use App\Models\Product;
+
 interface ProductRepositoryInterface
 {
     public function all();
-    public function store($data);
-    public function find($id);
-    public function update($data, $id);
-    public function destroy($id);
+    public function store(ProductData $data);
```
Comment (vladislavzubetcs) [inline app/Repositories/Interfaces/ProductRepositoryInterface.php]: empty space

File: app/Repositories/Interfaces/ProductRepositoryInterface.php [hunk 48]
```diff
@@ -1,12 +1,16 @@
...
+    public function update(Product $product , ProductData $data);
```
Comment (vladislavzubetcs) [inline app/Repositories/Interfaces/ProductRepositoryInterface.php]: add return types

File: app/Repositories/Interfaces/ServiceRepositoryInterface.php [hunk 49]
```diff
@@ -0,0 +1,16 @@
...
+declare(strict_types=1);
+
+namespace App\Repositories\Interfaces;
+
+use App\Data\ServiceData;
+use App\Models\Service;
+
+interface ServiceRepositoryInterface
+{
+    public function all();
```
Comment (vladislavzubetcs) [inline app/Repositories/Interfaces/ServiceRepositoryInterface.php]: same

**Review by vladislavzubetcs — CHANGES_REQUESTED** [review]

File: app/Data/ProductData.php (hunk 29 above)
Comment (dyamorph) [reply app/Data/ProductData.php:34]: fixed

File: app/Data/ServiceData.php (hunk 30 above)
Comment (dyamorph) [reply app/Data/ServiceData.php:25]: fixed

File: app/Enums/Currencies.php (hunk 31 above)
Comment (dyamorph) [reply app/Enums/Currencies.php:2]: fixed

File: app/Data/ProductData.php (hunk 32 above)
Comment (dyamorph) [reply app/Data/ProductData.php:2]: fixed

File: app/Data/ServiceData.php (hunk 33 above)
Comment (dyamorph) [reply app/Data/ServiceData.php:2]: fixed

File: app/Http/Controllers/ClientController.php (hunk 34 above)
Comment (dyamorph) [reply app/Http/Controllers/ClientController.php]: fixed

File: app/Http/Controllers/ClientController.php (hunk 35 above)
Comment (dyamorph) [reply app/Http/Controllers/ClientController.php]: fixed

File: app/Http/Controllers/ServiceController.php (hunk 38 above)
Comment (dyamorph) [reply app/Http/Controllers/ServiceController.php]: fixed

File: app/Http/Requests/Product/ProductStoreRequest.php (hunk 40 above)
Comment (dyamorph) [reply app/Http/Requests/Product/ProductStoreRequest.php]: fixed

File: app/Http/Requests/Product/ProductUpdateRequest.php (hunk 41 above)
Comment (dyamorph) [reply app/Http/Requests/Product/ProductUpdateRequest.php]: fixed

File: app/Models/Product.php (hunk 42 above)
Comment (dyamorph) [reply app/Models/Product.php:15]: fixed

File: app/Providers/RepositoryServiceProvider.php (hunk 44 above)
Comment (dyamorph) [reply app/Providers/RepositoryServiceProvider.php]: fixed

File: app/Providers/RepositoryServiceProvider.php (hunk 45 above)
Comment (dyamorph) [reply app/Providers/RepositoryServiceProvider.php]: fixed

File: app/Providers/RepositoryServiceProvider.php (hunk 46 above)
Comment (dyamorph) [reply app/Providers/RepositoryServiceProvider.php]: fixed

File: app/Repositories/Interfaces/ProductRepositoryInterface.php (hunk 47 above)
Comment (dyamorph) [reply app/Repositories/Interfaces/ProductRepositoryInterface.php]: fixed

File: app/Repositories/Interfaces/ProductRepositoryInterface.php (hunk 48 above)
Comment (dyamorph) [reply app/Repositories/Interfaces/ProductRepositoryInterface.php]: fixed

File: app/Repositories/Interfaces/ServiceRepositoryInterface.php (hunk 49 above)
Comment (dyamorph) [reply app/Repositories/Interfaces/ServiceRepositoryInterface.php]: fixed

File: app/Models/Product.php (hunk 43 above)
Comment (dyamorph) [reply app/Models/Product.php:28]: fixed

**Pull request merged**